tests/api2/test_virt_002_instance.py

```diff
@@ -0,0 +1,153 @@
+from threading import Event
+
+from middlewared.test.integration.assets.filesystem import mkfile
+from middlewared.test.integration.assets.pool import dataset
+from middlewared.test.integration.utils.client import client
+from middlewared.test.integration.utils.call import call
+from middlewared.test.integration.utils.ssh import ssh
+
+from auto_config import pool_name
+
+
+def clean():
+    call('virt.global.update', {'pool': ''}, job=True)
+    ssh(f'zfs destroy -r {pool_name}/.ix-virt || true')
+    call('virt.global.update', {'pool': 'tank'}, job=True)
+
+
+def test_virt_instances_create():
+    clean()
+
+    wait_agent = Event()
+
+    def wait_debian(*args, **kwargs):
+        wait_agent.set()
+
+    with client() as c:
+        c.subscribe('virt.instance.agent_running', wait_debian, sync=True)
+
+        # Create first so there is time for the agent to start
+        call('virt.instance.create', {'name': 'debian', 'image': 'debian/trixie', 'instance_type': 'VM'}, job=True)
+
+        call('virt.instance.create', {'name': 'void', 'image': 'voidlinux/musl'}, job=True)
+        ssh('incus exec void cat /etc/os-release | grep "Void Linux"')
+
+        call('virt.instance.create', {
+            'name': 'arch',
+            'image': 'archlinux/current/default',
+            'devices': [
+                {'name': 'tpm', 'dev_type': 'TPM', 'path': '/dev/tpm0', 'pathrm': '/dev/tmprm0'},
+            ],
+        }, job=True)
+        ssh('incus exec arch cat /etc/os-release | grep "Arch Linux"')
+
+        devices = call('virt.instance.device_list', 'arch')
+        assert any(i for i in devices if i['name'] == 'tpm'), devices
+
+        assert wait_agent.wait(timeout=30)
+        ssh('incus exec debian cat /etc/os-release | grep "Debian"')
+
+
+def test_virt_instances_update():
+    call('virt.instance.update', 'void', {'cpu': '1', 'memory': 500, 'environment': {'FOO': 'BAR'}}, job=True)
+    ssh('incus exec void grep MemTotal: /proc/meminfo|grep 512000')
+    # Checking CPUs seems to cause a racing condition (perhaps CPU currently in use in the container?)
+    # rv = ssh('incus exec void cat /proc/cpuinfo |grep processor|wc -l')
+    # assert rv.strip() == '1'
+    rv = ssh('incus exec void env | grep ^FOO=')
+    assert rv.strip() == 'FOO=BAR'
+
+
+def test_virt_instances_state():
+    # Stop only one of them so the others are stopped during delete
+    assert ssh('incus list void -f json| jq ".[].status"').strip() == '"Running"'
```

Comment:
Why are we using CLI only here?
Shouldn't we use the relevant middleware endpoint for this and for good measure perhaps use the CLI to double check of sorts?